### DISCUSSION TITLE
Integration with GNOME Files (aka Nautilus)

### DISCUSSION THREAD

**Discussion by ofseed** [ghostty-org/ghostty#2416] [Feature Requests, Ideas] [answered]
Ghostty has had a great user experience on Linux now, especially on the GNOME desktop, thanks to the utilization of libadwaita. I think it is even more native than the GNOME Console.

Nautilus has a right-click context menu item "Open in console," which is provided by the GNOME Console. Thus, you will end up opening the GNOME console no matter what default you set.

Are you interested in providing such a feature by Nautilus' C-bindings like what the GNOME Console did, so we could make Ghostty even more native?

**Comment by mitchellh** [comment]
I'm absolutely open to it. Can you provide more details what the integration looks like from both a UX and technical direction?

**Reply by ofseed** [reply]
![image](https://github.com/user-attachments/assets/a8d4aa12-6dbd-48b1-8074-9814863c3be3)

The UX is very simple, it's just a menu item in the right-click menu, click to open the terminal with the same working directory in Nautilus.

A similar usage is the "Send to device" provided by [GSConnect](https://github.com/GSConnect/gnome-shell-extension-gsconnect) as shown in the picture below. 

![image](https://github.com/user-attachments/assets/bc2bf559-5362-4d3f-b575-ea9763a77c55)

GSConnect's Nautilus extension needs the user to have Nautilus's Python bindings installed to work. It may be possible to use its C bindings by Zig so that we can provide it directly without any additional install requirements for the user. Here is the documentation: https://gnome.pages.gitlab.gnome.org/nautilus

I'm very willing to implement this feature, but I won't block anyone else who wants to work on it since I'm not that familiar with Zig.

**Comment by ofseed** [comment]
I just relaized that GNOME Console removed the extension and implemented this directly in Nautilus since GNOME 43. But it still a valid way to do this.

https://gitlab.gnome.org/GNOME/console/-/merge_requests/110

**Comment by tristan957** [comment]
I wonder if the most correct implementation would be to add an `Open in Terminal` option to the Nautilus menu that calls `xdg-terminal-exec`.

**Comment by jcollie** [comment]
`xdg-terminal-exec` is part of the solution - we need to configure `xdg-terminal-exec` to know about Ghostty, but we also need to finish the work to get Ghostty to open up new surfaces with a custom config. There's some basic plumbing in place but it needs to be exposed in the Zig API and then needs to be exposed via DBus or similar.

**Comment by ofseed** [comment]
`xdg-terminal-exec` is currently a community solution (and I'm using it), it has had a proposal MR in freedesktop's GitLab for five years already. But I'm hesitant on this too, I thought GNOME doesn't care about the community because there are many solutions other than writing kgx (aka GNOME Console) to nautilus's code directly. It might be possible to cause them change as long as the proposal not been closed.

**Comment by ofseed** [comment]
Another potential solution is to implement an open action like kitty. kitty provided a `.desktop` file for this action thus we can right-click the folder and open it by kitty

**Comment by 00-kat** [comment, marked as answer]
This exists now #4816.

**Comment by 2zqa** [comment]
In case this isn't working for anyone: you need to have the `nautilus-python` package installed.

**Reply by bittner** [reply]
@2zqa Can you confirm that you have a "Open in Ghostty" context menu entry in Nautilus after installing the `nautilus-python` package.

Is there anything else to do apart from installing the two packages? I install both `nautilus-python` and `ghostty` on NixOS (unstable, Xantusia, 25.11.20251002.7df7ff7), and nothing shows up in Nautilus.

Versions: Ghostty 1.2.0, GNOME nautilus 48.3, nautilus-python 4.0.1

**Reply by 2zqa** [reply]
Yes. I'm on NixOS stable. Config: https://github.com/2zqa/nix. I don't use Ghostty myself anymore (prefer kgx/GNOME Console), but my config  + ghostty and python-nautilus + rebooting was all I had to do

<img width="597" height="624" alt="Nautilus with right click menu showing 'Open in Ghostty'" src="https://github.com/user-attachments/assets/e29970af-3726-4278-8839-5698c8f53ce7" />

**Reply by musjj** [reply]
Can't get it to show up in Nautilus either on NixOS. You sure you didn't have `nautilus-open-any-terminal` or something similar?

**Reply by 2zqa** [reply]
I can still reproduce it working today with this config: https://github.com/2zqa/nix/commit/42d9e518ed615dca3458834b0e82f519249b4597

<img width="478" height="534" alt="afbeelding" src="https://github.com/user-attachments/assets/30bc9e04-d0be-4a68-9283-caac4080da92" />

**Reply by bittner** [reply]
Thanks @2zqa, that helped!

@musjj I can confirm this works, but only if you install both Ghostty and nautilus-python via systemPackages. Not with Home Manager, which is unfortunate, because I'd love to configure keyboard shortcuts. (We should find out why it doesn't work with Home Manager—and make it work!)

**Reply by bittner** [reply]
I found a workaround that allows to have the "Open in Ghostty" context menu option _and_ configure Ghostty using Home Manager:

1. Install the two system packages `ghostty` and `nautilus-python` (via `environment.systemPakages`).
2. Use the [programs.ghostty](https://mynixos.com/home-manager/options/programs.ghostty) Home Manager option to configure Ghostty, but set the [package option](https://mynixos.com/home-manager/option/programs.ghostty.package) to a dummy package (e.g. `pkgs.hello`), which avoids that "Open in Ghostty" appears twice in the context menu.

Not beautiful but does the job. :checkered_flag:

**Reply by musjj** [reply]
It looks like that the problem is with the `NAUTILUS_4_EXTENSION_DIR` variable. If you set it like this:

```nix
home.sessionVariables.NAUTILUS_4_EXTENSION_DIR="${pkgs.nautilus-python}/lib/nautilus/extensions-4";
```

It'll work even if you install both `ghostty` and `nautilus-python` from `home-manager`. I think this issue is relevant:

- https://github.com/NixOS/nixpkgs/issues/378139

**Reply by bittner** [reply]
<s>I don't know if this is a coincidence, but since I have Ghostty integrated in my Nautilus context menu I can't start it via the desktop shortcut anymore, neither from the dock nor from the GNOME application list (e.g. <kbd>🐧</kbd>+"Ghostty"). I now (have to) open Nautilus and use the context menu to start the terminal! :zany_face: </s>

**EDIT:** This problem was introduced by the Ghostty project. See https://github.com/ghostty-org/ghostty/discussions/10617 for details and a fix.

**Reply by bittner** [reply]
Just a heads-up: My final solution is to install the two system packages `ghostty` and `nautilus-python`, and then configure Ghostty simply by creating a `~/.config/ghostty/config` file, e.g.

```nix
  home-manager.users.${username}.home.xdg.configFile."ghostty/config".text = ''
    bell-features = system,audio,border,no-title
    font-size = 10
    keybind = ctrl+shift+arrow_left=goto_split:left
    keybind = ctrl+shift+arrow_right=goto_split:right
    keybind = ctrl+shift+arrow_up=goto_split:up
    keybind = ctrl+shift+arrow_down=goto_split:down
    window-height = 45
    window-width = 145
  '';
```
